### ISSUE TITLE
The garbage collector is not working correctly

### ISSUE THREAD

**Issue by objectProgrammer** [mirkosertic/Bytecoder#274]
I have a project that for unknown reasons stops working after about (2019-11-03 ::: 105 seconds; 2019-11-04 ::: 90 seconds). I don't know if this is caused by the garbage collector or not but because I have no way of debugging it I can't know whats causing it either. The error is `RuntimeError: index out of bounds` and I'd like further instructions on how to debug a wasm bytecoder program.

My guess is that it is caused by looping a list.

**Comment by mirkosertic** [comment]
Such problems are hard to resolve. Did you try compiling to `js`? If JS is working without problems, `Wasm` memory management is most likely the problem.

Here we have multiple options: did you try to increase the maximum page size? Maybe GC pressure is too high because of low free memory. If this does not work, we have to analyze the memory by hand. 

Maybe there is a self enclosing reference loop in the object structure. They are hard to detect and might confuse the GC,

**Comment by objectProgrammer** [comment]
I couldn't get the JS backend to work. Some events weren't firing, some mathematical expressions were resulting in incorrect results and other various bugs I have no idea how to deal with.
I increased wasmMaximumPages to 8192 but no difference was made.
Can you specify what a `self enclosing reference loop in the object structure` is? I have an AnimationFrameCallback which `requestAnimationFrame`s itself using `this`. I also tried referencing it from a static variable (Main: `static AnimationFrameCallback a = ...;` and then in the AnimationFrameCallback class: `(...).requestAnimationFrame(Main.a);`) but that made no difference. I also have a few HashMap loops (tried with both `for` and `forEach`, no difference) inside the AnimationFrameCallback called from the `run` method.

**Comment by mirkosertic** [comment]
I was refering to cycles in the object reference graph, which might confuse the GC.
Can you please try to log the current memory usage by calling the `freeMem` export every time the `requestAnimationFrame` callback runs? If the value steadily decreases, there is a memory leak for sure, maybe due to a bug in `de.mirkosertic.bytecoder.classlib.MemoryManager`

**Comment by objectProgrammer** [comment]
It indeed starts with about 32 millions and keeps decreasing and decreasing until it reaches 7604 (first test) and 10484 (second test) and then throws the error. I still have no idea to what `cycles in the object reference graph` is, maybe an example would help?

**Comment by mirkosertic** [comment]
A simple `cycle` is basically a member of object A referencing object B, and B is referencing A. More complex examples can be created by adding more objects.

Btw. do you have examples of `JS` backend not working? I'd love to fix those bugs.

**Comment by objectProgrammer** [comment]
The keydown event was not firing (couldn't reproduce in a new project) and some byte-to-int conversions were messed up (also couldn't reproduce in a new project but I was able to fix by using `Byte.toUnsignedInteger(...)`)

I'll check for `cycle`s in my code.

**Comment by objectProgrammer** [comment]
I indeed have a cycle (maybe more) in my project. I have one class where I store instances of classes and these (static) fields are then in turn accessed by other instances of classes also stored in the class. How should I fix this? Would the singleton pattern work? (Instead of `MyClass.someClass` I would use `SomeClass.getInstance()`)

**Comment by objectProgrammer** [comment]
I kind of liked that because I could then import static MyClass.*; and use `window` without doing Window.window() all the time.

**Comment by mirkosertic** [comment]
Can you please submit example code to show your pattern?

**Comment by objectProgrammer** [comment]
*com.example.Main*
```java
public class Main {

	static class MyAnimation implements AnimationFrameCallback {
		@Override
		public void run(int aElapsedTime) {
			Console.console().log(String.valueOf(MemoryManager.freeMem()));
			Window.window().requestAnimationFrame(this);
		}
	}

	public static void main(String[] args) {
		Window.window().requestAnimationFrame(new MyAnimation());
	}
}
```

*index.html*
```html
<meta charset="UTF-8">
<script src="bytecoder_wasmbindings.js"></script>
<script>
	var bytecoderWasmFile = 'bytecoder.wasm';
	var instantiated = function(result) {
		bytecoder.init(result.instance);
		bytecoder.exports.initMemory(0);
		bytecoder.exports.bootstrap(0);
		bytecoder.exports.main(0);
	};
	WebAssembly.instantiateStreaming(fetch(bytecoderWasmFile), bytecoder.imports).then(instantiated).catch(function(error) {
		var request = new XMLHttpRequest();
		request.open('GET', bytecoderWasmFile);
		request.responseType = 'arraybuffer';
		request.send();
		request.onload = function() {
			var bytes = request.response;
			WebAssembly.instantiate(bytes, bytecoder.imports).then(instantiated);
		};
	});
</script>
```

Bytecoder & BytecoderGradle 2019-11-04, `backend = wasm`, `mainClass = "com.example.Main"`, all other values default. It takes a long time but the value eventually reaches really low values and throws the error.

To speed up the process I tried `wasmMinimumPages = 64` and `wasmMaximumPages = 128` and it indeed did throw the error after a while.

This example cannot be related to cycles at all because as you can see it is very simple and only has 2 classes, 2 methods in total.

When I tried the JS backend I got an error (which I already had gotten earlier, but it was easy to fix) `TypeError: v is undefined` at `Window.window().requestAnimationFrame(this);`.

*Working version for JS backend com.example.Main*
```java
public class Main {
	static MyAnimation myAnimation;

	static class MyAnimation implements AnimationFrameCallback {
		@Override
		public void run(int aElapsedTime) {
			Console.console().log(String.valueOf(MemoryManager.freeMem()));
			Window.window().requestAnimationFrame(myAnimation);
		}
	}

	public static void main(String[] args) {
		Window.window().requestAnimationFrame(myAnimation = new MyAnimation());
	}
}
```

*js.html*
```html
<meta charset="UTF-8">
<script src="bytecoder.js"></script>
<script>
	bytecoder.bootstrap();
	bytecoder.exports.main();
</script>
```

Of course, because Bytecoder does not handle JS memory it always logs 0 but I think you might want to fix that you cannot reference `this` as a requestAnimationFrame target in AnimationFrameCallback.

**Comment by mirkosertic** [comment]
#279

**Comment by objectProgrammer** [comment]
Could you reproduce the issue I'm having?

**Comment by mirkosertic** [comment]
I can confirm a problem with GC. In fact GC currently runs never. However, the GC method is available as an export, but calling it by hand leads to an endless loop. I am investigating...

**Comment by objectProgrammer** [comment]
Thank you.

**Comment by mirkosertic** [comment]
Update: the internal memory management datastructures are fragmented over time.
Example: we allocate 100 bytes of data, and free it again. The freed block is moved from the allocated to the free memory list, and to manage this block info in the list, we need 8 bytes of data. So we have finally only 92 bytes remaining. This process continues over time. The result is a memory leak.

I am currently thinking about how to avoid this fragmentation...

**Comment by mirkosertic** [comment]
Good news, I found a solution for this bug. Will merge the fix to master as soon as travis-ci is working again. It currentl has some problems installing OpenJDK13 for the regression test suite.

**Comment by objectProgrammer** [comment]
That's great! Will you make a new release as well? I think this fix is important enough for a new release.

**Comment by mirkosertic** [comment]
A new release is synced to Maven Central and will be available soon.

Please note I had to modify the example a little bit:

Java:
```java
public class Main {
	static MyAnimation myAnimation;

	static class MyAnimation implements AnimationFrameCallback {
		@Override
		public void run(int aElapsedTime) {
			Console.console().log(String.valueOf((int) MemoryManager.freeMem())); // Cast to int
			Window.window().requestAnimationFrame(myAnimation);
		}
	}

	public static void main(String[] args) {
		Window.window().requestAnimationFrame(myAnimation = new MyAnimation());
	}
}
```

The memory size must be casted to an int due to the not complete support of long in JS and Wasm backend. Long to String conversion sometimes generates strange results.

Also please take a look at https://mirkosertic.github.io/Bytecoder/chapter-1/page-1-a/. I've posted some example code to enable the GC as part of the HTML embeddings.

**Comment by objectProgrammer** [comment]
Sorry to say, but I'm still having this problem. I copied the exact code you sent but it's still decreasing and decreasing (not even running the GC). Using Bytecoder 2019-11-25 with my Gradle plugin, if that matters.

**Comment by objectProgrammer** [comment]
And also, there was a little syntax error in your GC example code (fixed by #286).

**Comment by mirkosertic** [comment]
You have to enable the GC using the mentioned setTimeout JS code. Without this GC is inactive, and memory usage is increasing steadily.

**Comment by objectProgrammer** [comment]
Problem is that memory usage is decreasing and decreasing until it throws the exception, even without running the GC at all. Running the GC does not change anything.

**Comment by mirkosertic** [comment]
Please find attached a working demo with a running GC.

The console log shows a decreasing amount of free memory. Then you see the message "GC", and
the free memory bounces back to the initial amout, as memory was freed by the GC process.
I have run this demo for hours without problems.

[workinggcdemo.zip](https://github.com/mirkosertic/Bytecoder/files/3893283/workinggcdemo.zip)

**Comment by objectProgrammer** [comment]
Oh, I'm sorry. I was a little confused and for some reason I interpreted the number as the amount of memory currently in use, not only that but the wasm loader file was cached in my browser.

Sadly, the problems don't end here. In my testing project everything works fine but in my actual project the browser tab freezes when the GC runs. I tried decreasing the initial pages but that didn't change anything. It happens in both Firefox and Chromium and once again I don't have any idea how to debug it.
